### ISSUE TITLE
The handling of implicit arguments should be improved

### ISSUE THREAD

**Issue by mietek** [agda/agda#2099]
Filling either hole succeeds, but fails on subsequent reload.  This is Agda 2.5.1.1.

``` agda
module Regrets where

open import Data.Sum using (_⊎_ ; inj₁ ; inj₂ ; [_,_])
open import Data.Unit using (⊤ ; tt)

data Ty : Set where
  ι   : Ty
  _∨_ : Ty → Ty → Ty

record Model : Set₁ where
  field
    World : Set
    _≤_    : World → World → Set
    refl≤  : ∀ {w} → w ≤ w
    trans≤ : ∀ {w w′ w″} → w ≤ w′ → w′ ≤ w″ → w ≤ w″
    _‼_    : World → Ty → Set

open Model {{…}} public

module _ {{_ : Model}} where
  mutual
    _⊪_ : World → Ty → Set
    w ⊪ ι       = ⊤
    w ⊪ (A ∨ B) = (w ⊩ A) ⊎ (w ⊩ B)

    _⊩_ : World → Ty → Set
    w ⊩ A = ∀ {C w′} → w ≤ w′ → (∀ {w″} → w′ ≤ w″ → w″ ⊪ A → w″ ‼ C) → w′ ‼ C

  mutual
    mono⊩ : ∀ {A w w′} → w ≤ w′ → w ⊩ A → w′ ⊩ A
    mono⊩ ξ a = λ ξ′ k′ → a (trans≤ ξ ξ′) k′

    mono⊪ : ∀ {A w w′} → w ≤ w′ → w ⊪ A → w′ ⊪ A
    mono⊪ {ι}     ξ s = tt
    mono⊪ {A ∨ B} ξ s = [ (λ a → {!inj₁ (mono⊩ {A} ξ a)!})
                         , (λ b → {!inj₂ (mono⊩ {B} ξ b)!})
                         ] s
```

> Cannot instantiate the metavariable `_74` to solution `((_ Model.≤ .w)`
> `.w′₁ → ({w″ : Model.World _} → (_ Model.≤ .w′₁) w″ → w″ ⊪ A → (_`
> `Model.‼ w″) .C) → (_ Model.‼ .w′₁) .C)` since it contains the
> variable `.w′₁` which is not in scope of the metavariable or
> irrelevant in the metavariable but relevant in the solution
> when checking that the expression `a` has type
> `((_ Model.≤ .w) .w′₁ →`
> `({w″ : Model.World _} →`
> `(_ Model.≤ .w′₁) w″ → w″ ⊪ A → (_ Model.‼ w″) .C) →`
> `(_ Model.‼ .w′₁) .C)`
> 
> Lorem ipsum dolor `_75` sit amet `((_ Model.≤ .w)`
> `.w′₁ → ({w″ : Model.World _} → (_ Model.≤ .w′₁) w″ → w″ ⊪ B → (_`
> `Model.‼ w″) .C) → (_ Model.‼ .w′₁) .C)` consectetur adipiscing
> elit `.w′₁` sed do eiusmod tempor incididunt
> iut labore et dolore magna aliqua ut enim ad minim veniam
> quis nostrud exercitation ullamco laboris `b` nisi ut
> `((_ Model.≤ .w) .w′₁ →`
> `({w″ : Model.World _} →`
> `(_ Model.≤ .w′₁) w″ → w″ ⊪ B → (_ Model.‼ w″) .C) →`
> `(_ Model.‼ .w′₁) .C)`

**Comment by andreasabel** [comment]
The problem disappears without instance search:
```agda

record ⊤ : Set where
  constructor tt

data _⊎_ (A B : Set) : Set where
  inj₁ : A → A ⊎ B
  inj₂ : B → A ⊎ B

[_,_] : ∀{A B C : Set} (f : A → C) (g : B → C) → A ⊎ B → C
[ f , g ] (inj₁ x) = f x
[ f , g ] (inj₂ x) = g x

data Ty : Set where
  ι   : Ty
  _∨_ : Ty → Ty → Ty

record Model : Set₁ where
  field
    World : Set
    _≤_    : World → World → Set
    refl≤  : ∀ {w} → w ≤ w
    trans≤ : ∀ {w w′ w″} → w ≤ w′ → w′ ≤ w″ → w ≤ w″
    _‼_    : World → Ty → Set

-- open Model {{…}} public
-- module _ {{_ : Model}} where

module _ (M : Model) (open Model M) where
  mutual
    _⊪_ : World → Ty → Set
    w ⊪ ι       = ⊤
    w ⊪ (A ∨ B) = (w ⊩ A) ⊎ (w ⊩ B)

    _⊩_ : World → Ty → Set
    w ⊩ A = ∀ {C w′} → w ≤ w′ → (∀ {w″} → w′ ≤ w″ → w″ ⊪ A → w″ ‼ C) → w′ ‼ C

  mutual
    mono⊩ : ∀ {A w w′} → w ≤ w′ → w ⊩ A → w′ ⊩ A
    mono⊩ ξ a = λ ξ′ k′ → a (trans≤ ξ ξ′) k′

    mono⊪ : ∀ {A w w′} → w ≤ w′ → w ⊪ A → w′ ⊪ A
    mono⊪ {ι}     ξ s = tt
    mono⊪ {A ∨ B} ξ s = [ (λ a → {!inj₁ (mono⊩ {A} ξ a)!})
                         , (λ b → {!inj₂ (mono⊩ {B} ξ b)!})
                         ] s
```

**Comment by andreasabel** [comment]
The problem persists also without instance arguments.

``` agda
...
module _ (M : Model) (open Model M) where
  mutual
    _⊪_ : World → Ty → Set
    w ⊪ ι       = ⊤
    w ⊪ (A ∨ B) = (w ⊩ A) ⊎ (w ⊩ B)

    _⊩_ : World → Ty → Set
    w ⊩ A = ∀ {C w′} → w ≤ w′ → (∀ {w″} → w′ ≤ w″ → w″ ⊪ A → w″ ‼ C) → w′ ‼ C

  mutual
    mono⊩ : ∀ {A w w′} → w ≤ w′ → w ⊩ A → w′ ⊩ A
    mono⊩ ξ a = λ ξ′ k′ → a (trans≤ ξ ξ′) k′

    mono⊪ : ∀ {A w w′} → w ≤ w′ → w ⊪ A → w′ ⊪ A
    mono⊪ {ι}     ξ s = tt
    mono⊪ {A ∨ B} ξ s = [ (λ a →  inj₁ (mono⊩ {A} ξ a) )
                        , (λ b →  inj₂ (mono⊩ {B} ξ b) ) ] s
```

**Comment by andreasabel** [comment]
This is the old problem with eager insertion of implicit arguments.  If `mono⊩ {A} ξ a` is a hole, it's type can be inferred, but if you reload, it is expanded too far to `mono⊩ {A} ξ a {_} {_}`.  Make `C` visible and your plight disappears:

``` agda
...
    _⊩_ : World → Ty → Set
    w ⊩ A = ∀ C {w′} → w ≤ w′ → (∀ {w″} → w′ ≤ w″ → w″ ⊪ A → w″ ‼ C) → w′ ‼ C

  mutual
    mono⊩ : ∀ {A w w′} → w ≤ w′ → w ⊩ A → w′ ⊩ A
    mono⊩ ξ a = λ _ ξ′ → a _ (trans≤ ξ ξ′)

    mono⊪ : ∀ {A w w′} → w ≤ w′ → w ⊪ A → w′ ⊪ A
    mono⊪ {A ∨ B} ξ = [ (λ a →  inj₁ (mono⊩ {A} ξ a) )
                      , (λ b →  inj₂ (mono⊩ {B} ξ b) )
                      ]
```

Related issues are #1079 and #1095.

**Comment by acertain** [comment]
Here's another, maybe more minimal example of a related problem:

```agda
module implicit where

the : (A : Set) (x : A) -> A
the _ x = x

data It {X : Set}(x : X) : Set where it : It x

data Wrap (x : Set) : Set where wrap : (a : x) -> Wrap x

prf : {X Y : Set} -> Wrap ({a : X} (y : X) -> It a)
prf {X} {Y} = wrap (the {!{a : X} (y : X) -> It a!} (\y -> it))
```

It typechecks with the hole filled, but you can't fill the hole.

The problem is you don't know if you need to insert implicit abstractions until the hole is filled. 

How to implement a typechecker that doesn't have this problem? I'd expect the changes to agda to be quite invasive, but is it worth it to fix this with future languages? Does it add too much complexity and slow down typechecking too much to be worth it?

**Comment by andreasabel** [comment]
In this particular case, for the type checking problem
```agda
prf : {X Y : Set} → Wrap ({a : X} (y : X) → It a)
prf {X} = wrap (the ? (λ y → it))
```
There are two possible hidden lambda insertions,
```agda
prf {X} = wrap (λ {a} → the ? (λ y → it))
```
and
```agda
prf {X} = wrap (the ? (λ {a} y → it))
```
Agda picks the first because `the ...` is an application. To say "picks" gives the wrong impression, because Agda eagerly inserts the hidden lambda there and never tries to get a picture of the alternatives.
The use of `the ?` prevents Agda to push the type in to the lambda expression `\ y -> it`, thus, it is facing the problem `\ y -> it : ?` and at this point there is a heuristics to speculate `? = (y : ?1) -> ?2 y`.
If you solve the `?` yourself, then Agda has the target type of the lambda and inserts a hidden abstraction, ending up with:
```agda
elaboration : {X Y : Set} → Wrap ({a : X} (y : X) → It a)
elaboration {X} = wrap (λ {a} → the ({a : X} (y : X) → It a) (λ {a} y → it) {a})
```

In general, the problem is undecidable, because Curry-style System F is undecidable already.
The radical solution is to get rid of first-class hidden quantification and only allow argument hiding on the syntax (scope checker) level.  That was proposed by Conor McBride, and was originally like this in Coq, but users like the flexibility of the hiding approach in agda.

What would help is to let the user declare if a hidden argument is to be inserted eagerly (whenever possible) or lazily (only if absolutely forced).  This is possible in Coq, and Lean even has two notations for eager vs. lazy hidden quantification.

Finally, there is even a Master thesis on this issue, see http://www2.tcs.ifi.lmu.de/~abel/MScThesisJohanssonLloyd.pdf . However, the solution proposed there was not entirely convincing.

**Comment by jespercockx** [comment]
This is more of a research problem than a concrete proposal, so I'm putting it in the `icebox`.

**Comment by jespercockx** [comment]
It seems @AndrasKovacs has a proposal on how to fix this problem at https://github.com/AndrasKovacs/elaboration-zoo/tree/master/experimental/poly-instantiation.

**Comment by nad** [comment]
@AndrasKovacs, how is the example described by @andreasabel [above](https://github.com/agda/agda/issues/2099#issuecomment-305958549) handled by your system?

**Comment by mietek** [comment]
Additional test cases can be found in these issues:
https://github.com/agda/agda/issues/1079
https://github.com/agda/agda/issues/1092
https://github.com/agda/agda/issues/1095
https://github.com/agda/agda/issues/1533
https://github.com/agda/agda/issues/2412
https://github.com/agda/agda/issues/2500
https://github.com/agda/agda/issues/2766
https://github.com/agda/agda/issues/3468
https://github.com/agda/agda/issues/3559

These issues are potentially related:
https://github.com/agda/agda/issues/1387
https://github.com/agda/agda/issues/2428
https://github.com/agda/agda/issues/2556

**Comment by AndrasKovacs** [comment]
@mietek Thanks for gathering this, I'll review the examples.

@nad Regarding that example:

First, I did a bit more simplification, and looked at the Church-encoded version of the following code in my implementation:

```
{-# OPTIONS --type-in-type #-}

data Wrap (A : Set) : Set where
  wrap : A → Wrap A

the : (A : Set) → A → A
the = λ A x → x

prf : Wrap ({A : Set} → A → A)
prf = wrap (the ? (λ x → x))
```
Observations:

#### Behavior of my current implementation

It does the same thing as Agda: it chooses the outer quantifier instead of the inner one. With holes, we would get the same phenomenon as in Agda. That's because this example is not just about unknown lambda insertion, it also features unknown implicit argument insertion: if we choose the inner quantification, we need an implicit application, but we don't need it otherwise. 

The outer quantification:
```
wrap (λ {A} → the (A → A) (λ (x : A) → x))
```
The inner one:
```
wrap (λ {A} → the (∀{A} → A → A) (λ {A}(x : A) → x){A})
```

I currently *do not* insert fresh telescope applications (i.e. applications with undetermined arity) into the elaboration output, because these often result in unification problems where a meta is applied to an unknown telescope, such as

```
α{β σ : γ δ`} =? rhs
```
where `α`, `β` and `γ` are all metas, `α` is applied to the `β σ` telescope term
which has type `γ δ`, and `σ` and `δ` are spines.

One example, which seems to pop up often:

```
α{β A : γ A} =? A
```
A sensible solution to this problem would be:

```
γ = λ A. ((A : U) ▶ ∙)
β = λ A. (A ∷ [])
α = λ {A}. A
```
But this is clearly outside the usual pattern fragment, and we'd need to investigate more which equations with unknown spine applications have sensible solutions. So for the time being I'm focusing on what we can do only with lambda insertions, which appear much easier. 

#### Behavior with telescope application insertions

I also checked this case - it is a minor code change, but we get a number of regressions because of
the lack of worked-out meta-solving strategies with telescopes in spines. Here, when we infer a type for
```
the ? (λ x → x)
```
we get a telescope function type, because `(λ x → x)` is checked with an unknown  type, hence we wrap it in a telescope lambda, and thus we get a telescope function as the overall inferred type. Then, we insert a telescope application after `(λ x → x)`:
```
the ? (λ {g : γ A} (x : α A g) → x){β A:γ A}
```
In this case, it seems that we get the correct behavior.

There are two valid elaboration outputs (inner/outer quantification), both of which can be possibly obtained from filling the hole, and which are not definitionally equal. Hence my algorithm produces an unsolved constraint. The unsolved constraint looks like this:

```
α A {β A : γ A} =? A
```
Here we can either solve as 
```
  γ = λ A. ∙
  α = λ A. A
```
or as
```
  γ = λ A. (A : U) ▶ ∙ 
  β = λ A. A
  α = λ _ {A}. A
```
The first one is the outer quantification, the second is the inner one. We would be able to fill the hole in both ways without getting a type error, and in both cases the unsolved constraint would be solved. In contrast, Agda commits to the outer quantification prematurely, and runs into an error when it turns out to be the wrong choice. 

#### In summary
- I currently have a simpler algorithm which only inserts telescope lambdas, and which is probably conservative over Agda, in that it accepts strictly more programs, and seems to cover many problematic cases. I will have a better idea about this after I check the test cases collated by mietek. 
- There is a more substantial possible extension to unification which handles unknown telescopes in meta spines, and which allows inserted telescope applications. We can simply do the telescope application insertions without upgrading unification, but that results in some regressions.

**Comment by nad** [comment]
Here is another example:
```agda
postulate
  A : Set
  P : A → Set

data Kind : Set where
  implication : Kind

_↝[_]_ : Set → Kind → Set → Set
A ↝[ implication ] B = A → B

id : ∀ {k} {A B : Set} → A ↝[ k ] B → A ↝[ k ] B
id f = f

works : (∀ x → P x) → {x : A} → P x → P x
works = id (λ hyp {_} _ → hyp _)

fails : (∀ x → P x) → {x : A} → P x → P x
fails = id (λ hyp _ → hyp _)
```
Does your system handle this example correctly?

**Comment by AndrasKovacs** [comment]
@nad: I can't reproduce that closely because it uses constructor-headed function inference. But just looking at it, it would not work, because I don't sandwich telescopes between lambdas. I would check `(λ hyp _ → hyp _)` with the meta-headed type `α ↝[ β ] γ `, which results in a telescope lambda inserted only on the outside of the expression. Possible solution would be more aggressive telescope insertion, perhaps it would work well, I haven't yet checked. Good point though!

I'll post tomorrow a review of the other related code examples.

**Comment by AndrasKovacs** [comment]
I added extra telescope lambda insertion in order to support @nad's previous example.

I looked at the code examples which were sufficiently reproducible. [Here's a link](https://github.com/AndrasKovacs/implicit-fun-elaboration/blob/8df450d530da5f674485c81f6d5704f072c08285/fcif/Agdaissues.txt) to the file that I produced, containing some discussion.

In short:

- https://github.com/agda/agda/issues/2099#issuecomment-305922361 : would be fixed on a hypothetical extension of my current system.
- https://github.com/agda/agda/issues/2099#issuecomment-540799683 : would be fixed on my system, it is currently not representable because of lacking features but analogous cases are representable.
-  https://github.com/agda/agda/issues/1079#issuecomment-129028215 : is not fixed on my system, I believe the issue is all about eta conversion for Bool, which issue cannot be avoided even with the fanciest elaborator for implicit functions.
- https://github.com/agda/agda/issues/1079#issuecomment-392325549 : is fixed on my system.
- https://github.com/agda/agda/issues/1079#issuecomment-393411912 : is fixed on my system, but this should be solvable just by not skipping lambda insertion for local let-s, no fancy inference needed.
- https://github.com/agda/agda/issues/3468#issue-393677107 : could be in principle fixed on my system, but I prefer it non-fixed, because it involves inferring a non-dependent implicit function, which my implementation explicitly excludes (because such functions are useless practically).
- https://github.com/agda/agda/issues/1387#issue-99824785 : would be fixed on my system, it does not work right now because of lack of general constraint postponing.

**Comment by AndrasKovacs** [comment]
Update & question. First, my proposal is now explained in more detail in paper form: https://dl.acm.org/doi/10.1145/3408983

Question: I'm somewhat mystified as to what happens in Agda 2.6.1 and later:

```
open import Data.Maybe

IdTy = {A : Set} → A → A

test1 : Maybe IdTy     -- works
test1 = just (λ x → x)

test2 : (M : Set₁ → Set)(j : ∀{A} → A → M A) → M IdTy
test2 M j = j (λ x → x) -- fails

postulate
  M : Set₁ → Set

test3 : (j : ∀{A} → A → M A) → M IdTy
test3 j = j (λ x → x) -- works

test4 : (j : ∀{A : Set₁} → A → Maybe A) → Maybe IdTy
test4 j = j (λ x → x) -- works

test5 : (j : ∀{A : Set₁} → A → Maybe A) → Maybe IdTy
test5 j = let x = j (λ x → x) in x -- fails
```

From `test1` and `test2` I would deduce that the following happens: there is a rule for checking inductive constructor-headed expressions with their inductive type-constructor-headed types. In that case, we can start by instantiating all the type parameters, since they can be always recovered from the type. By the time we check `λ x → x` in `test1`, the parameter type is already known, so we're good. `test2` fails because no inductive parameters are involved, only arbitrary bound vars with given types. 

But I get from `test3-test5` that something else/more must happen. It seems that there's some more general magic involved when checking with a postulated or inductive type constructor head. From `test5` I assume that the magic only works in checking mode.
